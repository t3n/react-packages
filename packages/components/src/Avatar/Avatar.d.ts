import React, { ReactNode } from 'react';
import { TextColorProps } from 'styled-system';
export interface AvatarImageProps {
    optimizeSrc?: boolean;
    src?: string;
    size?: number;
    className?: string;
    alt?: string;
    loading?: boolean;
    children?: ReactNode;
}
export interface AvatarProps extends Omit<AvatarImageProps, 'className'> {
    label?: string;
    textColor?: TextColorProps['color'];
}
declare const Avatar: React.FC<AvatarProps>;
export default Avatar;
