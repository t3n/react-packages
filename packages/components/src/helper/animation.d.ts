import { ThemeProps } from '@t3n/theme';
export declare const backgroundAnimation: import("styled-components").FlattenInterpolation<import("styled-components").ThemedStyledProps<ThemeProps, any>>;
