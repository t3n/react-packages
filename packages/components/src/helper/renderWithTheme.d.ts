import 'jest-styled-components';
import React from 'react';
import { RenderOptions } from '@testing-library/react';
import { Theme } from '@t3n/theme';
interface OptionalTheme {
    theme?: Theme;
}
export declare const renderWithTheme: (ui: React.ReactElement, { theme, ...options }: RenderOptions & OptionalTheme) => import("@testing-library/react").RenderResult<typeof import("@testing-library/dom/types/queries"), HTMLElement>;
export {};
