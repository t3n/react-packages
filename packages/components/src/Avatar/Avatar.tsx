/* eslint-disable react/require-default-props */
import React, { ReactNode } from 'react';
import styled from 'styled-components';
import {
  border,
  color,
  color as styledColor,
  space,
  TextColorProps,
} from 'styled-system';

import Box from '../Box';
import Image from '../Image';
import Placeholder from '../Placeholder';
import Text from '../Text';

export interface AvatarImageProps {
  optimizeSrc?: boolean;
  src?: string;
  size?: number;
  className?: string;
  alt?: string;
  loading?: boolean;
  children?: ReactNode;
}

export interface AvatarProps extends Omit<AvatarImageProps, 'className'> {
  label?: string;
  textColor?: TextColorProps['color'];
}

const AvatarPlaceholder = styled((props) => (
  <Placeholder as="span" {...props} />
))`
  ${({ theme }) =>
    border({
      theme,
      borderWidth: 2,
      borderColor: 'shades.white',
      borderStyle: 'solid',
      borderRadius: '50%',
    })};
`;

const DefaultAvatar = styled((props) => <Box as="span" {...props} />)`
  ${({ theme }) =>
    border({
      theme,
      borderWidth: 2,
      borderColor: 'shades.white',
      borderStyle: 'solid',
      borderRadius: '50%',
    })};

  ${({ theme }) =>
    color({
      theme,
      bg: 'brand.red',
      color: 'text.inverse',
    })};
`;

const StyledText = styled((props) => <Text {...props} />)`
  user-select: none;
  text-transform: uppercase;
`;

const AvatarImage = ({
  src,
  optimizeSrc = true,
  size = 40,
  className,
  alt,
  loading,
}: AvatarImageProps) => {
  const parts = alt?.split(' ').filter(Boolean);
  const initials =
    parts && parts.length
      ? `${parts[0][0]}${parts.length > 1 ? parts[parts.length - 1][0] : ''}`
      : '?';

  if (loading)
    return <AvatarPlaceholder height={`${size}px`} width={`${size}px`} />;

  if (!src)
    return (
      <DefaultAvatar
        display="flex"
        alignItems="center"
        height={`${size}px`}
        width={`${size}px`}
      >
        {alt ? (
          <StyledText bold p={1} mx="auto" my={0} small>
            {initials}
          </StyledText>
        ) : (
          <StyledText bold p={1} mx="auto" my={0} small>
            ?
          </StyledText>
        )}
      </DefaultAvatar>
    );

  if (optimizeSrc)
    return (
      <Image
        src={src}
        className={className}
        width={size}
        height={size}
        alt={alt || 'Avatar'}
      />
    );

  return (
    <img
      src={src}
      className={className}
      width={size}
      height={size}
      alt={alt || 'Kein Avatar'}
    />
  );
};

const StyledAvatarImage = styled(AvatarImage)<{ src?: string }>`
  border-radius: 50%;
  border: 2px solid white;
  box-sizing: border-box;
`;

const StyledAvatar = styled.span`
  display: flex;
  align-items: center;
`;

const AvatarLabel = styled.span<Pick<AvatarProps, 'textColor'>>`
  ${({ theme }) => space({ pl: 2, theme })}
  ${({ textColor, theme }) => styledColor({ color: textColor, theme })}
`;

const Avatar: React.FC<AvatarProps> = ({
  label = '',
  textColor,
  children,
  ...rest
}) => (
  <StyledAvatar>
    <StyledAvatarImage {...rest} />
    {label && label.length > 0 && (
      <AvatarLabel textColor={textColor}>{label}</AvatarLabel>
    )}
    {children}
  </StyledAvatar>
);

export default Avatar;
