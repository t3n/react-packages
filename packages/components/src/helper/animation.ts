import { css, keyframes } from 'styled-components';

import { ThemeProps } from '@t3n/theme';

const backgroundAnimationKeyframes = ({ theme }: ThemeProps) => keyframes`
  0% {
    background-color: ${theme.colors.shades.grey204}
  }

  50% {
    background-color: ${theme.colors.shades.grey244}
  }

  100% {
    background-color: ${theme.colors.shades.grey204}
  }
`;

// eslint-disable-next-line import/prefer-default-export
export const backgroundAnimation = css`
  animation: ${backgroundAnimationKeyframes} 2s linear infinite;
`;
