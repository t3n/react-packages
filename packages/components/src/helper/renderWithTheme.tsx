/* eslint-disable import/no-extraneous-dependencies */
import 'jest-styled-components';

import React from 'react';
import { render, RenderOptions } from '@testing-library/react';
import { ThemeProvider } from 'styled-components';

import { Theme, theme as originalTheme } from '@t3n/theme';

interface OptionalTheme {
  theme?: Theme;
}

// eslint-disable-next-line import/prefer-default-export
export const renderWithTheme = (
  ui: React.ReactElement,
  { theme, ...options }: RenderOptions & OptionalTheme
) => {
  // merge over default theme

  const mergedTheme = { ...originalTheme, ...theme };

  const Wrapper: React.FC = ({ children }) => (
    <ThemeProvider theme={mergedTheme}>{children}</ThemeProvider>
  );

  return render(ui, { wrapper: Wrapper, ...options });
};
