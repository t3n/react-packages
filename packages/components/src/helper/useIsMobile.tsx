import { Theme } from '@t3n/theme';
import { useCallback, useContext, useEffect, useState } from 'react';
import { ThemeContext } from 'styled-components';

const useIsMobile = (): boolean => {
  const [isMobile, setIsMobile] = useState(false);

  const theme: Theme = useContext(ThemeContext);
  const fontSize = window
    .getComputedStyle(window.document.body)
    .getPropertyValue('font-size');
  const mobileBreakpoint =
    parseInt(theme.breakpoints[1], 10) * parseInt(fontSize, 10);

  const testIsMobile = useCallback(() => {
    if (window.matchMedia(`(max-width: ${mobileBreakpoint}px)`).matches) {
      setIsMobile(true);
    } else {
      setIsMobile(false);
    }
  }, [mobileBreakpoint]);

  useEffect(() => {
    testIsMobile();
    window.addEventListener('resize', testIsMobile);

    return () => {
      window.removeEventListener('resize', testIsMobile);
    };
  }, [testIsMobile]);

  return isMobile;
};

export default useIsMobile;
