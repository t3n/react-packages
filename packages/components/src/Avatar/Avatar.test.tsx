import React from 'react';

import { renderWithTheme } from '../helper/renderWithTheme';
import Avatar from './Avatar';

test('Avatar matches snapshot and renders label', () => {
  const { container, getByText } = renderWithTheme(
    <Avatar size={40} src="" label="Some author" alt="alt-attribute" />,
    {},
  );
  expect(container.firstChild).toMatchSnapshot();
  expect(getByText(/some author/i)).toBeDefined();
});

test('Avatar renders without label', () => {
  const { container } = renderWithTheme(
    <Avatar size={40} src="" alt="alt-attribute" />,
    {},
  );

  expect(container.firstChild).toMatchSnapshot();
});
