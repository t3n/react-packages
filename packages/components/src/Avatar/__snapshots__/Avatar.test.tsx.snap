// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Avatar matches snapshot and renders label 1`] = `
.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  height: 40px;
  width: 40px;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c3 {
  font-family: -apple-system,system-UI,BlinkMacSystemFont,Segoe UI,Roboto,Helvetica Neue,Arial,sans-serif;
  font-size: 0.875rem;
  line-height: 1.5em;
  font-weight: bold;
  padding: 4px;
  margin-left: auto;
  margin-right: auto;
  margin-top: 0;
  margin-bottom: 0;
}

.c2 {
  border-width: 2px;
  border-color: #fff;
  border-style: solid;
  border-radius: 50%;
  background-color: #f9423a;
  color: #fff;
}

.c4 {
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  text-transform: uppercase;
}

.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c5 {
  padding-left: 8px;
}

<div
  class="c0"
>
  <div
    class="c1 c2"
    display="flex"
    height="40px"
    width="40px"
  >
    <p
      class="c3 c4"
    >
      a
    </p>
  </div>
  <span
    class="c5"
  >
    Some author
  </span>
</div>
`;

exports[`Avatar renders without label 1`] = `
.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  height: 40px;
  width: 40px;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c3 {
  font-family: -apple-system,system-UI,BlinkMacSystemFont,Segoe UI,Roboto,Helvetica Neue,Arial,sans-serif;
  font-size: 0.875rem;
  line-height: 1.5em;
  font-weight: bold;
  padding: 4px;
  margin-left: auto;
  margin-right: auto;
  margin-top: 0;
  margin-bottom: 0;
}

.c2 {
  border-width: 2px;
  border-color: #fff;
  border-style: solid;
  border-radius: 50%;
  background-color: #f9423a;
  color: #fff;
}

.c4 {
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  text-transform: uppercase;
}

.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

<div
  class="c0"
>
  <div
    class="c1 c2"
    display="flex"
    height="40px"
    width="40px"
  >
    <p
      class="c3 c4"
    >
      a
    </p>
  </div>
</div>
`;
